### ISSUE TITLE
Once Fips is enable can't encrypt/decrypt file on version 3.0.2-fips

### ISSUE THREAD

**Issue by zivhadad24** [openssl/openssl#24962]
I have installed ubuntu22.04 with Fips enabled
OpenSSL version is: OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)

I'm trying to encrypt file with this command **openssl enc -aes-256-cbc -iter 600000 -in {file} -out {output file} -pass pass:{passphrase}**
And getting the following error

```
**BIO[0x55c5c451b5e0]: write(0,8) - FILE pointer
BIO[0x55c5c451b5e0]: write return 1 processed: 8
BIO[0x55c5c451b5e0]: write(0,8) - FILE pointer
BIO[0x55c5c451b5e0]: write return 1 processed: 8
PKCS5_PBKDF2_HMAC failed
80FBA20ACB7F0000:error:1C800070:Provider routines:kdf_pbkdf2_set_ctx_params:invalid salt length:../providers/implementations/kdfs/pbkdf2.c:233:
BIO[0x55c5c451b2a0]: Free - FILE pointer
BIO[0x55c5c451b5e0]: Free - FILE pointer**
```

I have read your doc and saw that once -iter is specified the default salt length should be 16bytes 
But actually once running the command we can see that it writes only 8 bytes 

Since it generates only 8 bytes for salt it's failed on FIps checks which needs a least 16 bytes of salt
I see that in any version that bigger then OpenSSL 3.2 you have added the -saltlen option
So also if I'm trying to encrypt with version OpenSSL 3.2 and then decrypt on version 3.0.2 I'm getting the same error 
Since the code on 3.0.2 is trying to find 8 bytes of salt which is hardcode in the C code

**Comment by nhorman** [comment]
Can you try with openssl 3.3.0?  I just attempted it and it seems to work fine there, so it appears we've fixed something that was deemed non-backportable

**Comment by zivhadad24** [comment]
@nhorman Hi thank you for answering
I already checked 3.3.0 it works fine
The issue is that in Ubuntu22.04 the stable release of openssl is 3.0.2 the issue is started only once FIPS-MODULE is enabled 
OpenSSL 3.0.2 should be fips compatible but it doesn't since the default Salt Length is 8 bytes but once you enable FIPS some checks are in place including the checking if the slat length is not less then 16 bytes 
OpenSSL 3.0.2 Generates random salt with limit of 8 bytes and then checks if the salt is not less then 16 bytes which is something the never can be true 

As I see the option of specifying Salt length is available only on versions that equal or Greater  then 3.2.0  
But what about who is using Ubuntu22.04 and there is no option to upgrade ?

**Comment by nhorman** [comment]
the saltlen option was created here https://github.com/openssl/openssl/pull/21858, and the decision was made to not backport it.

If you need to use FIPS on version 3.0.2, It appears that you can do so without using the -iter option, which , from the look of the master branch will disable the use of pbkdf2, falling back to an older kdf that should work in that environment

**Comment by zivhadad24** [comment]
@nhorman But the requirements for FIPS is to use pbkdf2 kdf any other are not certified 
And the requirement is also 16 Bytes of Saltlength
Even if not using pbkdf2 there is no option to use Saltlength greater then 8 bytes since it's hardcoded 
 # define PKCS5_SALT_LEN                  8  compiled time 

This means that any version of SSL that is less then 3.2.0 can't be FIPS certified since the ask for a least 16 bytes of slat

**Comment by nhorman** [comment]
The requirement that we assert is in our security policy, see page 22 of:
https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp4282.pdf

The enc applet falls back to using a key based derivation function using the provided cipher (aes-128-cbc) and the hash function sha256.

**Comment by zivhadad24** [comment]
@nhorman So do you mean that once the fallback is coming in 
There will not be any salt mechanism involved and that okay for FIPS requirements ?

**Comment by nhorman** [comment]
No, I mean that the fallback is to use a salt mechanism based on SHA256 and aes-128.  See enc_main around line 578:

```
} else {
                BIO_printf(bio_err, "*** WARNING : "
                                    "deprecated key derivation used.\n"
                                    "Using -iter or -pbkdf2 would be better.\n");
                if (!EVP_BytesToKey(cipher, dgst, sptr,
                                    (unsigned char *)str, str_len,
                                    1, key, iv)) {
                    BIO_printf(bio_err, "EVP_BytesToKey failed\n");
                    goto end;
                }
            } 
```

**Comment by xnox** [comment]
Are you using Ubuntu Pro FIPS, or self built OpenSSL FIPS module? As the source code and behaviour of the two are widely different.

**Comment by ryankenney-dev** [comment]
I can confirm the same behavior with Ubuntu Pro FIPS:
```
... Provider routines:kdf_pbkdf2_set_ctx_params:invalid salt length:../providers/implementations/kdfs/pbkdf2.c:233 ...
```

Packages installed:
```
openssl - 3.0.2-0ubuntu1.19+Fips1
openssl-fips-module-3:amd64 - 3.0.5-0ubuntu0.1+Fips2.1
```

Runtime version:
```
openssl version
OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)
````

No custom openssl builds.

**Comment by xnox** [comment]
> I can confirm the same behavior with Ubuntu Pro FIPS:
> 
> ```
> ... Provider routines:kdf_pbkdf2_set_ctx_params:invalid salt length:../providers/implementations/kdfs/pbkdf2.c:233 ...
> ```
> 
> Packages installed:
> 
> ```
> openssl - 3.0.2-0ubuntu1.19+Fips1
> openssl-fips-module-3:amd64 - 3.0.5-0ubuntu0.1+Fips2.1
> ```
> 
> Runtime version:
> 
> ```
> openssl version
> OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)
> ```
> 
> No custom openssl builds.

Please seek support from Ubuntu pro.

**Comment by xnox** [comment]
Please also note https://github.com/openssl/openssl/pull/27446 which might be related.
